<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Working with digital pathology data and python | The Surgical Informatics Cookbook</title>
  <meta name="description" content="All the brilliant bits of code we don’t want to forget." />
  <meta name="generator" content="bookdown 0.28 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Working with digital pathology data and python | The Surgical Informatics Cookbook" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="All the brilliant bits of code we don’t want to forget." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Working with digital pathology data and python | The Surgical Informatics Cookbook" />
  
  <meta name="twitter:description" content="All the brilliant bits of code we don’t want to forget." />
  

<meta name="author" content="Surgical Informatics, University of Edinburgh" />


<meta name="date" content="2023-08-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="redcap.html"/>
<link rel="next" href="citations-with-r-markdown.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">The Surgical Informatics Cookbook</a></li>

<li class="divider"></li>
<li><a href="index.html#section" id="toc-section"></a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#how-to-contribute"><i class="fa fa-check"></i><b>1.1</b> How to contribute</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#indexing"><i class="fa fa-check"></i><b>1.2</b> Indexing</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="introduction.html"><a href="introduction.html#index"><i class="fa fa-check"></i><b>1.2.1</b> Index</a></li>
<li class="chapter" data-level="1.2.2" data-path="introduction.html"><a href="introduction.html#chapter-and-section-references"><i class="fa fa-check"></i><b>1.2.2</b> Chapter and section references</a></li>
<li class="chapter" data-level="1.2.3" data-path="introduction.html"><a href="introduction.html#figure-and-table-references"><i class="fa fa-check"></i><b>1.2.3</b> Figure and table references</a></li>
<li class="chapter" data-level="1.2.4" data-path="introduction.html"><a href="introduction.html#citations"><i class="fa fa-check"></i><b>1.2.4</b> Citations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="snippets.html"><a href="snippets.html"><i class="fa fa-check"></i><b>2</b> Snippets</a>
<ul>
<li class="chapter" data-level="2.1" data-path="snippets.html"><a href="snippets.html#useful-rstudio-functions"><i class="fa fa-check"></i><b>2.1</b> Useful RStudio functions</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="snippets.html"><a href="snippets.html#clean-restart"><i class="fa fa-check"></i><b>2.1.1</b> Clean restart</a></li>
<li class="chapter" data-level="2.1.2" data-path="snippets.html"><a href="snippets.html#remove-duplicated-packages-between-personal-and-system-libraries"><i class="fa fa-check"></i><b>2.1.2</b> Remove duplicated packages between personal and system libraries</a></li>
<li class="chapter" data-level="2.1.3" data-path="snippets.html"><a href="snippets.html#remove-large-file-accidentally-committed-to-git"><i class="fa fa-check"></i><b>2.1.3</b> Remove large file accidentally committed to git</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="snippets.html"><a href="snippets.html#memory-and-efficient-ram-usage"><i class="fa fa-check"></i><b>2.2</b> Memory and efficient RAM usage</a></li>
<li class="chapter" data-level="2.3" data-path="snippets.html"><a href="snippets.html#exporting-tables-that-work-in-both-pdf-and-word"><i class="fa fa-check"></i><b>2.3</b> Exporting tables that work in both PDF and Word</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="snippets.html"><a href="snippets.html#warning-in-kableextra.-please-specify-format-in-kable.-kableextra-can-customize-either-html-or-latex-outputs."><i class="fa fa-check"></i><b>2.3.1</b> Warning in kableExtra. Please specify format in kable. kableExtra can customize either HTML or LaTeX outputs.</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="snippets.html"><a href="snippets.html#creating-reproducible-r-examples-to-share-in-the-group-binder-holepunch-and-docker"><i class="fa fa-check"></i><b>2.4</b> Creating Reproducible R Examples to Share in the Group (binder, holepunch and docker)</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="snippets.html"><a href="snippets.html#create-basic-reproducible-examples"><i class="fa fa-check"></i><b>2.4.1</b> Create Basic Reproducible Examples</a></li>
<li class="chapter" data-level="2.4.2" data-path="snippets.html"><a href="snippets.html#holepunch---complex-reproducible-examples"><i class="fa fa-check"></i><b>2.4.2</b> <code>holepunch</code> - Complex Reproducible Examples</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="snippets.html"><a href="snippets.html#working-with-chis"><i class="fa fa-check"></i><b>2.5</b> Working with CHIs</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="snippets.html"><a href="snippets.html#chi_dob---extract-date-of-birth-from-chi"><i class="fa fa-check"></i><b>2.5.1</b> <code>chi_dob()</code> - Extract date of birth from CHI</a></li>
<li class="chapter" data-level="2.5.2" data-path="snippets.html"><a href="snippets.html#chi_gender---extract-gender-from-chi"><i class="fa fa-check"></i><b>2.5.2</b> <code>chi_gender()</code> - Extract gender from CHI</a></li>
<li class="chapter" data-level="2.5.3" data-path="snippets.html"><a href="snippets.html#chi_age---extract-age-from-chi"><i class="fa fa-check"></i><b>2.5.3</b> <code>chi_age()</code> - Extract age from CHI</a></li>
<li class="chapter" data-level="2.5.4" data-path="snippets.html"><a href="snippets.html#chi_valid---logical-test-for-valid-chi"><i class="fa fa-check"></i><b>2.5.4</b> <code>chi_valid()</code> - Logical test for valid CHI</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="snippets.html"><a href="snippets.html#working-with-dates"><i class="fa fa-check"></i><b>2.6</b> Working with dates</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="snippets.html"><a href="snippets.html#difference-between-two-dates"><i class="fa fa-check"></i><b>2.6.1</b> Difference between two dates</a></li>
<li class="chapter" data-level="2.6.2" data-path="snippets.html"><a href="snippets.html#lags"><i class="fa fa-check"></i><b>2.6.2</b> Lags</a></li>
<li class="chapter" data-level="2.6.3" data-path="snippets.html"><a href="snippets.html#pulling-out-change-in-status-data"><i class="fa fa-check"></i><b>2.6.3</b> Pulling out “change in status” data</a></li>
<li class="chapter" data-level="2.6.4" data-path="snippets.html"><a href="snippets.html#guess-the-format-of-dates"><i class="fa fa-check"></i><b>2.6.4</b> Guess the Format of Dates</a></li>
<li class="chapter" data-level="2.6.5" data-path="snippets.html"><a href="snippets.html#isoweek-and-epiweek"><i class="fa fa-check"></i><b>2.6.5</b> Isoweek and epiweek</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="snippets.html"><a href="snippets.html#labelling-tables-and-figures-in-markdown-chunks"><i class="fa fa-check"></i><b>2.7</b> Labelling tables and figures in markdown chunks</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="snippets.html"><a href="snippets.html#plots"><i class="fa fa-check"></i><b>2.7.1</b> Plots</a></li>
<li class="chapter" data-level="2.7.2" data-path="snippets.html"><a href="snippets.html#tables"><i class="fa fa-check"></i><b>2.7.2</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="snippets.html"><a href="snippets.html#join-tables-and-overwrite-values-that-are-often-missing"><i class="fa fa-check"></i><b>2.8</b> Join tables and overwrite values (that are often missing)</a></li>
<li class="chapter" data-level="2.9" data-path="snippets.html"><a href="snippets.html#read-in-multiple-files"><i class="fa fa-check"></i><b>2.9</b> Read in multiple files</a></li>
<li class="chapter" data-level="2.10" data-path="snippets.html"><a href="snippets.html#averaging-multiple-aucs-say-from-imputed-datasets"><i class="fa fa-check"></i><b>2.10</b> Averaging multiple AUCs (say from imputed datasets)</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="data-manipulation.html"><a href="data-manipulation.html"><i class="fa fa-check"></i><b>3</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="data-manipulation.html"><a href="data-manipulation.html#collapse-multiple-no-and-yes-options"><i class="fa fa-check"></i><b>3.1</b> Collapse multiple “no” and “yes” options</a></li>
<li class="chapter" data-level="3.2" data-path="data-manipulation.html"><a href="data-manipulation.html#filtering-best-practice"><i class="fa fa-check"></i><b>3.2</b> Filtering best practice</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="data-manipulation.html"><a href="data-manipulation.html#from-jamie-farrell"><i class="fa fa-check"></i><b>3.2.1</b> From Jamie Farrell</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="data-manipulation.html"><a href="data-manipulation.html#filter-na-dropping-rows-where-all-specified-variables-are-na"><i class="fa fa-check"></i><b>3.3</b> Filter NA: Dropping rows where all specified variables are NA</a></li>
<li class="chapter" data-level="3.4" data-path="data-manipulation.html"><a href="data-manipulation.html#vectorising-rowwise-procedures"><i class="fa fa-check"></i><b>3.4</b> Vectorising rowwise procedures</a></li>
<li class="chapter" data-level="3.5" data-path="data-manipulation.html"><a href="data-manipulation.html#multiple-imputation-and-ipw-for-missing-data"><i class="fa fa-check"></i><b>3.5</b> Multiple imputation and IPW for missing data</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="data-manipulation.html"><a href="data-manipulation.html#data"><i class="fa fa-check"></i><b>3.5.1</b> Data</a></li>
<li class="chapter" data-level="3.5.2" data-path="data-manipulation.html"><a href="data-manipulation.html#multiple-imputation"><i class="fa fa-check"></i><b>3.5.2</b> Multiple imputation</a></li>
<li class="chapter" data-level="3.5.3" data-path="data-manipulation.html"><a href="data-manipulation.html#ipw"><i class="fa fa-check"></i><b>3.5.3</b> IPW</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="machine-learning.html"><a href="machine-learning.html"><i class="fa fa-check"></i><b>4</b> Machine learning</a>
<ul>
<li class="chapter" data-level="4.1" data-path="machine-learning.html"><a href="machine-learning.html#deep-learning"><i class="fa fa-check"></i><b>4.1</b> Deep learning</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="machine-learning.html"><a href="machine-learning.html#pulling-images-from-redcap-directly-to-argodeep"><i class="fa fa-check"></i><b>4.1.1</b> Pulling images from REDCap directly to argodeep</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html"><i class="fa fa-check"></i><b>5</b> Data Transfer and Eddie</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#uploading-and-downloading-data-the-easy-way"><i class="fa fa-check"></i><b>5.1</b> Uploading and Downloading Data the Easy Way</a></li>
<li class="chapter" data-level="5.2" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#alternative-methods-when-the-easy-way-wont-work"><i class="fa fa-check"></i><b>5.2</b> Alternative Methods when the Easy Way won’t work</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#what-is-eddie"><i class="fa fa-check"></i><b>5.2.1</b> What is Eddie?</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#using-the-command-line"><i class="fa fa-check"></i><b>5.3</b> Using the Command Line</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#downloads-and-setup---eddie-example"><i class="fa fa-check"></i><b>5.3.1</b> Downloads and Setup - Eddie Example</a></li>
<li class="chapter" data-level="5.3.2" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#copy-and-paste-with-putty"><i class="fa fa-check"></i><b>5.3.2</b> Copy and Paste with PuTTY</a></li>
<li class="chapter" data-level="5.3.3" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#closing-a-session-in-putty"><i class="fa fa-check"></i><b>5.3.3</b> Closing a Session in PuTTY</a></li>
<li class="chapter" data-level="5.3.4" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#eddie-file-structure"><i class="fa fa-check"></i><b>5.3.4</b> Eddie File Structure</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#copying-data-from-eddie-into-rstudio-server"><i class="fa fa-check"></i><b>5.4</b> Copying Data from Eddie into RStudio server</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#method-1-using-putty-or-another-terminalshell-connected-to-eddie"><i class="fa fa-check"></i><b>5.4.1</b> Method 1: Using PuTTY (or another terminal/shell connected to Eddie)</a></li>
<li class="chapter" data-level="5.4.2" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#method-2-using-the-rstudio-server-terminal"><i class="fa fa-check"></i><b>5.4.2</b> Method 2: Using the RStudio Server Terminal</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#the-surgical-informatics-group-folder-on-eddie"><i class="fa fa-check"></i><b>5.5</b> The Surgical Informatics Group Folder (On Eddie)</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#the-scratch-space"><i class="fa fa-check"></i><b>5.5.1</b> The Scratch Space</a></li>
<li class="chapter" data-level="5.5.2" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#other-spaces"><i class="fa fa-check"></i><b>5.5.2</b> Other Spaces</a></li>
<li class="chapter" data-level="5.5.3" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#running-eddie-from-rstudio-server"><i class="fa fa-check"></i><b>5.5.3</b> Running Eddie from RStudio Server</a></li>
<li class="chapter" data-level="5.5.4" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#modules-applications-in-eddie"><i class="fa fa-check"></i><b>5.5.4</b> Modules (Applications) in Eddie</a></li>
<li class="chapter" data-level="5.5.5" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#working-with-r-in-eddie"><i class="fa fa-check"></i><b>5.5.5</b> Working with R in Eddie</a></li>
<li class="chapter" data-level="5.5.6" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#specific-r-package-versions"><i class="fa fa-check"></i><b>5.5.6</b> Specific R Package Versions</a></li>
<li class="chapter" data-level="5.5.7" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#general-eddie-etiquette"><i class="fa fa-check"></i><b>5.5.7</b> General Eddie Etiquette</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#additional-resources"><i class="fa fa-check"></i><b>5.6</b> Additional Resources</a></li>
<li class="chapter" data-level="5.7" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#using-rstudio-directly-in-eddie"><i class="fa fa-check"></i><b>5.7</b> Using RStudio directly in Eddie</a></li>
<li class="chapter" data-level="5.8" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#configuring-python-deep-learning-environment-in-eddie"><i class="fa fa-check"></i><b>5.8</b> Configuring Python Deep Learning Environment in Eddie</a>
<ul>
<li class="chapter" data-level="5.8.1" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#package-installation"><i class="fa fa-check"></i><b>5.8.1</b> Package installation</a></li>
<li class="chapter" data-level="5.8.2" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#gpu-usage"><i class="fa fa-check"></i><b>5.8.2</b> GPU usage</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="data-transfer-and-eddie.html"><a href="data-transfer-and-eddie.html#configuring-jupyter-notebook-access"><i class="fa fa-check"></i><b>5.9</b> Configuring Jupyter Notebook Access</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="plotting.html"><a href="plotting.html"><i class="fa fa-check"></i><b>6</b> Plotting</a>
<ul>
<li class="chapter" data-level="6.1" data-path="plotting.html"><a href="plotting.html#gghighlight-example"><i class="fa fa-check"></i><b>6.1</b> GGHighlight Example</a></li>
<li class="chapter" data-level="6.2" data-path="plotting.html"><a href="plotting.html#coord_flip-factor-orders"><i class="fa fa-check"></i><b>6.2</b> <code>coord_flip()</code> factor orders</a></li>
<li class="chapter" data-level="6.3" data-path="plotting.html"><a href="plotting.html#axis-font-size"><i class="fa fa-check"></i><b>6.3</b> Axis font size</a></li>
<li class="chapter" data-level="6.4" data-path="plotting.html"><a href="plotting.html#shorten-arrows-on-a-dag"><i class="fa fa-check"></i><b>6.4</b> Shorten Arrows on a DAG</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="genomics.html"><a href="genomics.html"><i class="fa fa-check"></i><b>7</b> Genomics</a>
<ul>
<li class="chapter" data-level="7.1" data-path="genomics.html"><a href="genomics.html#single-cell-analysis"><i class="fa fa-check"></i><b>7.1</b> Single Cell Analysis</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="genomics.html"><a href="genomics.html#minimising-the-size-of-a-seurat-object"><i class="fa fa-check"></i><b>7.1.1</b> Minimising the size of a Seurat Object</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="genomics.html"><a href="genomics.html#nextflow"><i class="fa fa-check"></i><b>7.2</b> Nextflow</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="genomics.html"><a href="genomics.html#rnaseq-example-with-sbatch-slurm"><i class="fa fa-check"></i><b>7.2.1</b> RNASeq example with sbatch (slurm)</a></li>
<li class="chapter" data-level="7.2.2" data-path="genomics.html"><a href="genomics.html#dnaseq-example-with-sbatch-slurm"><i class="fa fa-check"></i><b>7.2.2</b> DNASeq example with sbatch (slurm)</a></li>
<li class="chapter" data-level="7.2.3" data-path="genomics.html"><a href="genomics.html#atacseq-example-with-sbatch-slurm"><i class="fa fa-check"></i><b>7.2.3</b> ATACSeq example with sbatch (slurm)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="programming-in-rlang.html"><a href="programming-in-rlang.html"><i class="fa fa-check"></i><b>8</b> Programming in rlang</a>
<ul>
<li class="chapter" data-level="8.1" data-path="programming-in-rlang.html"><a href="programming-in-rlang.html#rlang"><i class="fa fa-check"></i><b>8.1</b> rlang</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="programming-in-rlang.html"><a href="programming-in-rlang.html#what-is-rlang"><i class="fa fa-check"></i><b>8.1.1</b> What is rlang?</a></li>
<li class="chapter" data-level="8.1.2" data-path="programming-in-rlang.html"><a href="programming-in-rlang.html#dynamic-calling-of-variables-using-dplyr"><i class="fa fa-check"></i><b>8.1.2</b> Dynamic calling of variables using dplyr</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="server-admin.html"><a href="server-admin.html"><i class="fa fa-check"></i><b>9</b> Server admin</a>
<ul>
<li class="chapter" data-level="9.1" data-path="server-admin.html"><a href="server-admin.html#rstudio-argonaut-and-argosafe"><i class="fa fa-check"></i><b>9.1</b> RStudio (argonaut and argosafe)</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="server-admin.html"><a href="server-admin.html#scheduled-script-cron"><i class="fa fa-check"></i><b>9.1.1</b> Scheduled script (CRON)</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="server-admin.html"><a href="server-admin.html#rstudio-connect-argoshare"><i class="fa fa-check"></i><b>9.2</b> RStudio Connect: argoshare</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="server-admin.html"><a href="server-admin.html#delete-user"><i class="fa fa-check"></i><b>9.2.1</b> Delete user</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="server-admin.html"><a href="server-admin.html#shiny-server-opensource-setup---uoe-eleanor"><i class="fa fa-check"></i><b>9.3</b> Shiny server opensource setup - UoE Eleanor</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="server-admin.html"><a href="server-admin.html#setting-up-ubuntu-instance"><i class="fa fa-check"></i><b>9.3.1</b> Setting up Ubuntu Instance</a></li>
<li class="chapter" data-level="9.3.2" data-path="server-admin.html"><a href="server-admin.html#install-nginx"><i class="fa fa-check"></i><b>9.3.2</b> Install Nginx</a></li>
<li class="chapter" data-level="9.3.3" data-path="server-admin.html"><a href="server-admin.html#installing-r"><i class="fa fa-check"></i><b>9.3.3</b> Installing R</a></li>
<li class="chapter" data-level="9.3.4" data-path="server-admin.html"><a href="server-admin.html#installing-shiny-server"><i class="fa fa-check"></i><b>9.3.4</b> Installing Shiny Server</a></li>
<li class="chapter" data-level="9.3.5" data-path="server-admin.html"><a href="server-admin.html#installing-other-linux-packages"><i class="fa fa-check"></i><b>9.3.5</b> Installing other Linux packages</a></li>
<li class="chapter" data-level="9.3.6" data-path="server-admin.html"><a href="server-admin.html#installing-r-packages"><i class="fa fa-check"></i><b>9.3.6</b> Installing R packages</a></li>
<li class="chapter" data-level="9.3.7" data-path="server-admin.html"><a href="server-admin.html#loading-your-app"><i class="fa fa-check"></i><b>9.3.7</b> Loading your app</a></li>
<li class="chapter" data-level="9.3.8" data-path="server-admin.html"><a href="server-admin.html#make-your-app-the-default-index-page"><i class="fa fa-check"></i><b>9.3.8</b> Make your app the default index page</a></li>
<li class="chapter" data-level="9.3.9" data-path="server-admin.html"><a href="server-admin.html#getting-a-user-friendly-domain-name"><i class="fa fa-check"></i><b>9.3.9</b> Getting a user-friendly domain name</a></li>
<li class="chapter" data-level="9.3.10" data-path="server-admin.html"><a href="server-admin.html#appendix-1-password-protecting-your-site"><i class="fa fa-check"></i><b>9.3.10</b> Appendix 1 – Password protecting your site</a></li>
<li class="chapter" data-level="9.3.11" data-path="server-admin.html"><a href="server-admin.html#appendix-2-setting-up-a-swap-file"><i class="fa fa-check"></i><b>9.3.11</b> APPENDIX 2 – Setting up a swap file</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="redcap.html"><a href="redcap.html"><i class="fa fa-check"></i><b>10</b> REDCap</a>
<ul>
<li class="chapter" data-level="10.1" data-path="redcap.html"><a href="redcap.html#api-pull"><i class="fa fa-check"></i><b>10.1</b> API pull</a></li>
<li class="chapter" data-level="10.2" data-path="redcap.html"><a href="redcap.html#applying-a-redcap-factoring-script"><i class="fa fa-check"></i><b>10.2</b> Applying a REDCap factoring script</a></li>
<li class="chapter" data-level="10.3" data-path="redcap.html"><a href="redcap.html#alternative-to-the-factoring-script-redcap_label-from-librarycollaboratorr"><i class="fa fa-check"></i><b>10.3</b> Alternative to the factoring script: <code>redcap_label()</code> from <code>library(collaboratorR)</code></a></li>
<li class="chapter" data-level="10.4" data-path="redcap.html"><a href="redcap.html#scannable-barcodes-in-redcap"><i class="fa fa-check"></i><b>10.4</b> Scannable barcodes in REDCap</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html"><i class="fa fa-check"></i><b>11</b> Working with digital pathology data and python</a>
<ul>
<li class="chapter" data-level="11.1" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#installing-openslide"><i class="fa fa-check"></i><b>11.1</b> Installing OpenSlide</a></li>
<li class="chapter" data-level="11.2" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#using-openslide"><i class="fa fa-check"></i><b>11.2</b> Using OpenSlide</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#opening-a-slide-and-inspecting-properties"><i class="fa fa-check"></i><b>11.2.1</b> Opening a slide and inspecting properties</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#whole-workflow"><i class="fa fa-check"></i><b>11.3</b> Whole workflow</a></li>
<li class="chapter" data-level="11.4" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#using-python-and-r-together"><i class="fa fa-check"></i><b>11.4</b> Using python and R together</a></li>
<li class="chapter" data-level="11.5" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#handling-errors-in-python"><i class="fa fa-check"></i><b>11.5</b> Handling errors in python</a></li>
<li class="chapter" data-level="11.6" data-path="working-with-digital-pathology-data-and-python.html"><a href="working-with-digital-pathology-data-and-python.html#iterating-over-many-many-slide-files"><i class="fa fa-check"></i><b>11.6</b> Iterating over many many slide files</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html"><i class="fa fa-check"></i><b>12</b> Citations with R Markdown</a>
<ul>
<li class="chapter" data-level="12.1" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#zotero-set-up"><i class="fa fa-check"></i><b>12.1</b> Zotero Set-up</a></li>
<li class="chapter" data-level="12.2" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#file-storage-set-up"><i class="fa fa-check"></i><b>12.2</b> File Storage Set-up</a></li>
<li class="chapter" data-level="12.3" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#folder-set-up"><i class="fa fa-check"></i><b>12.3</b> Folder Set-up</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#storing-pdfs"><i class="fa fa-check"></i><b>12.3.1</b> Storing PDFs</a></li>
<li class="chapter" data-level="12.3.2" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#storing-bibliographies"><i class="fa fa-check"></i><b>12.3.2</b> Storing Bibliographies</a></li>
<li class="chapter" data-level="12.3.3" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#subfolder-for-references"><i class="fa fa-check"></i><b>12.3.3</b> Subfolder for References</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#configuring-zotero"><i class="fa fa-check"></i><b>12.4</b> Configuring Zotero</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#zotfile-pdf-preferences"><i class="fa fa-check"></i><b>12.4.1</b> ZotFile PDF Preferences</a></li>
<li class="chapter" data-level="12.4.2" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#general-zotero-settings"><i class="fa fa-check"></i><b>12.4.2</b> General Zotero Settings</a></li>
<li class="chapter" data-level="12.4.3" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#refresh-bibtex-citation-key"><i class="fa fa-check"></i><b>12.4.3</b> Refresh BibTex Citation Key</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#generating-a-.bib-file"><i class="fa fa-check"></i><b>12.5</b> Generating a .bib File</a></li>
<li class="chapter" data-level="12.6" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#linking-dropbox-and-rstudio-server"><i class="fa fa-check"></i><b>12.6</b> Linking DropBox and Rstudio Server</a>
<ul>
<li class="chapter" data-level="12.6.1" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#token-creation"><i class="fa fa-check"></i><b>12.6.1</b> Token Creation</a></li>
<li class="chapter" data-level="12.6.2" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#dropbox-linkage"><i class="fa fa-check"></i><b>12.6.2</b> DropBox Linkage</a></li>
</ul></li>
<li class="chapter" data-level="12.7" data-path="citations-with-r-markdown.html"><a href="citations-with-r-markdown.html#citing-r-packages"><i class="fa fa-check"></i><b>12.7</b> Citing R Packages</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">The Surgical Informatics Cookbook</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="working-with-digital-pathology-data-and-python" class="section level1 hasAnchor" number="11">
<h1><span class="header-section-number">Chapter 11</span> Working with digital pathology data and python<a href="working-with-digital-pathology-data-and-python.html#working-with-digital-pathology-data-and-python" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Histopathology is used to analyse tissues from the body under the microscope. When a sample of tissue is taken during and operation or biopsy, it is put into a preservative, dehydrated and then impregnated with paraffin wax. It may also be frozen. This allows the tissue to be very thinly sliced and mounted onto glass slides, where the tissue can then be stained and studied under the microscope. Different types of stains can be used to look at different things.</p>
<p>These stained slides can be scanned with a slide scanner to allow the image to be digitised using scanners. There are two main types of images. Brightfield - generated by using white light, which requires deconvolution to produce separate colour channels. Fluorescence - generated using a light source to excite fluorescent probes, this typically generates images with 2+ colour channels.</p>
<p>The file formats produced by these scanners varies by manufacturer and scanning technique. Essentially, data is stored as tiled TIFF images within what is possibly an XML structure. There are usually multiple images produced, not only of the tissue but also of the slide, including a macro image (no magnification, sometimes helps scanner locate tissue), and the slide label.</p>
<p>Unfortunately, it’s quite a difficult format to interrogate easily. However, there is a nice program written that can solve this issue, known as <a href="https://openslide.org/">OpenSlide</a>.</p>
<p>OpenSlide has C, Python and Java versions. The python version is probably easiest to interact with, either directly in Python, or it can be combined with R using <code>reticulate</code>.</p>
<p>This chapter will not go into depth about issues around environments or how to install python properly. Please familiarise yourself with how to use <code>pip</code>, particularly <code>pip install</code> and set / activate python environments.</p>
<p>You can use RStudio for working in python too and this is what is suggested here. Other formats exist including jupyter notebooks.</p>
<p>For more information, <a href="https://github.com/bnsreenu/python_for_microscopists">Python for microscopists</a> has excellent resources.</p>
<div id="installing-openslide" class="section level2 hasAnchor" number="11.1">
<h2><span class="header-section-number">11.1</span> Installing OpenSlide<a href="working-with-digital-pathology-data-and-python.html#installing-openslide" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>First install the linux version of openslide - the python files act as a wrapper to this system library.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb172-1"><a href="working-with-digital-pathology-data-and-python.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="ex">apt-get</span> install python3-openslide</span></code></pre></div>
<p>To install openslide and the libraries we will need, first create a python environment. Do this inside the folder you are going to be working in.</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb173-1"><a href="working-with-digital-pathology-data-and-python.html#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="ex">virtualenv</span> <span class="at">-p</span> python3 openslide_env</span></code></pre></div>
<p>You can then either activate the virtual environment to install python libraries there, or install them system/user wide.</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb174-1"><a href="working-with-digital-pathology-data-and-python.html#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python3</span> <span class="at">-m</span> pip install openslide-python</span></code></pre></div>
<p>We will also need some other libraries</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb175-1"><a href="working-with-digital-pathology-data-and-python.html#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python3</span> <span class="at">-m</span> pip install numpy matplotlib tifffile</span></code></pre></div>
<p>Installation is fairly straighforward -if there are issues, do it as a sudo user. Most of the problems will appear if you are trying to use different or conflicting versions of python or if there are incorrect permissions for installation.</p>
</div>
<div id="using-openslide" class="section level2 hasAnchor" number="11.2">
<h2><span class="header-section-number">11.2</span> Using OpenSlide<a href="working-with-digital-pathology-data-and-python.html#using-openslide" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Once openslide is installed, we can now write python scripts. Do this in RStudio or other IDE in the same folder as your virtual environment. You can create a project in RStudio using this existing folder, then add new python scripts. Python scripts have the suffix <code>.py</code>.</p>
<p>First, start by checking you can successfully import the libraries.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb176-1"><a href="working-with-digital-pathology-data-and-python.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> openslide <span class="im">import</span> open_slide</span>
<span id="cb176-2"><a href="working-with-digital-pathology-data-and-python.html#cb176-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> openslide</span>
<span id="cb176-3"><a href="working-with-digital-pathology-data-and-python.html#cb176-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> PIL <span class="im">import</span> Image</span>
<span id="cb176-4"><a href="working-with-digital-pathology-data-and-python.html#cb176-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb176-5"><a href="working-with-digital-pathology-data-and-python.html#cb176-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb176-6"><a href="working-with-digital-pathology-data-and-python.html#cb176-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tifffile <span class="im">as</span> tiff</span></code></pre></div>
<p>If you get an error- try installing the libraries with <code>pip</code> as above.</p>
<div id="opening-a-slide-and-inspecting-properties" class="section level3 hasAnchor" number="11.2.1">
<h3><span class="header-section-number">11.2.1</span> Opening a slide and inspecting properties<a href="working-with-digital-pathology-data-and-python.html#opening-a-slide-and-inspecting-properties" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb177"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb177-1"><a href="working-with-digital-pathology-data-and-python.html#cb177-1" aria-hidden="true" tabindex="-1"></a>file_path <span class="op">=</span> <span class="st">&#39;/home/test/slides/myslide.svs&#39;</span> <span class="co"># replace with where your slides are</span></span>
<span id="cb177-2"><a href="working-with-digital-pathology-data-and-python.html#cb177-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-3"><a href="working-with-digital-pathology-data-and-python.html#cb177-3" aria-hidden="true" tabindex="-1"></a>slide_in <span class="op">=</span> open_slide(file_path)</span>
<span id="cb177-4"><a href="working-with-digital-pathology-data-and-python.html#cb177-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-5"><a href="working-with-digital-pathology-data-and-python.html#cb177-5" aria-hidden="true" tabindex="-1"></a>slide_in_props <span class="op">=</span> slide_in.properties</span>
<span id="cb177-6"><a href="working-with-digital-pathology-data-and-python.html#cb177-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-7"><a href="working-with-digital-pathology-data-and-python.html#cb177-7" aria-hidden="true" tabindex="-1"></a>slide_in_assc_images <span class="op">=</span> slide_in.associated_images.items() <span class="co"># see if there are any associated images</span></span></code></pre></div>
</div>
</div>
<div id="whole-workflow" class="section level2 hasAnchor" number="11.3">
<h2><span class="header-section-number">11.3</span> Whole workflow<a href="working-with-digital-pathology-data-and-python.html#whole-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here is the whole workflow for extracting magnified tiles of the image, taken from <a href="https://github.com/bnsreenu/python_for_microscopists">Python for microscopists</a>.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb178-1"><a href="working-with-digital-pathology-data-and-python.html#cb178-1" aria-hidden="true" tabindex="-1"></a>file_path <span class="op">=</span> <span class="st">&#39;/home/test/slide_ins/myslide_in.svs&#39;</span> <span class="co"># replace with where your slide_ins are</span></span>
<span id="cb178-2"><a href="working-with-digital-pathology-data-and-python.html#cb178-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-3"><a href="working-with-digital-pathology-data-and-python.html#cb178-3" aria-hidden="true" tabindex="-1"></a>slide_in <span class="op">=</span> open_slide(file_path)</span>
<span id="cb178-4"><a href="working-with-digital-pathology-data-and-python.html#cb178-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-5"><a href="working-with-digital-pathology-data-and-python.html#cb178-5" aria-hidden="true" tabindex="-1"></a>slide_in_props <span class="op">=</span> slide_in.properties</span>
<span id="cb178-6"><a href="working-with-digital-pathology-data-and-python.html#cb178-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-7"><a href="working-with-digital-pathology-data-and-python.html#cb178-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(slide_in_props)</span>
<span id="cb178-8"><a href="working-with-digital-pathology-data-and-python.html#cb178-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-9"><a href="working-with-digital-pathology-data-and-python.html#cb178-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Vendor is:&quot;</span>, slide_in_props[<span class="st">&#39;openslide.vendor&#39;</span>])</span>
<span id="cb178-10"><a href="working-with-digital-pathology-data-and-python.html#cb178-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Pixel size of X in um is:&quot;</span>, slide_in_props[<span class="st">&#39;openslide.mpp-x&#39;</span>])</span>
<span id="cb178-11"><a href="working-with-digital-pathology-data-and-python.html#cb178-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Pixel size of Y in um is:&quot;</span>, slide_in_props[<span class="st">&#39;openslide.mpp-y&#39;</span>])</span>
<span id="cb178-12"><a href="working-with-digital-pathology-data-and-python.html#cb178-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-13"><a href="working-with-digital-pathology-data-and-python.html#cb178-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Objective used to capture the image</span></span>
<span id="cb178-14"><a href="working-with-digital-pathology-data-and-python.html#cb178-14" aria-hidden="true" tabindex="-1"></a>objective <span class="op">=</span> <span class="bu">float</span>(slide_in.properties[openslide.PROPERTY_NAME_OBJECTIVE_POWER])</span>
<span id="cb178-15"><a href="working-with-digital-pathology-data-and-python.html#cb178-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;The objective power is: &quot;</span>, objective)</span>
<span id="cb178-16"><a href="working-with-digital-pathology-data-and-python.html#cb178-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-17"><a href="working-with-digital-pathology-data-and-python.html#cb178-17" aria-hidden="true" tabindex="-1"></a><span class="co"># get slide_in dimensions for the level 0 - max resolution level</span></span>
<span id="cb178-18"><a href="working-with-digital-pathology-data-and-python.html#cb178-18" aria-hidden="true" tabindex="-1"></a>slide_in_dims <span class="op">=</span> slide_in.dimensions</span>
<span id="cb178-19"><a href="working-with-digital-pathology-data-and-python.html#cb178-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(slide_in_dims)</span>
<span id="cb178-20"><a href="working-with-digital-pathology-data-and-python.html#cb178-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-21"><a href="working-with-digital-pathology-data-and-python.html#cb178-21" aria-hidden="true" tabindex="-1"></a><span class="co">#Get a thumbnail of the image and visualize</span></span>
<span id="cb178-22"><a href="working-with-digital-pathology-data-and-python.html#cb178-22" aria-hidden="true" tabindex="-1"></a>slide_in_thumb_600 <span class="op">=</span> slide_in.get_thumbnail(size<span class="op">=</span>(<span class="dv">600</span>, <span class="dv">600</span>))</span>
<span id="cb178-23"><a href="working-with-digital-pathology-data-and-python.html#cb178-23" aria-hidden="true" tabindex="-1"></a>slide_in_thumb_600.show()</span>
<span id="cb178-24"><a href="working-with-digital-pathology-data-and-python.html#cb178-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-25"><a href="working-with-digital-pathology-data-and-python.html#cb178-25" aria-hidden="true" tabindex="-1"></a><span class="co">#Convert thumbnail to numpy array</span></span>
<span id="cb178-26"><a href="working-with-digital-pathology-data-and-python.html#cb178-26" aria-hidden="true" tabindex="-1"></a>slide_in_thumb_600_np <span class="op">=</span> np.array(slide_in_thumb_600)</span>
<span id="cb178-27"><a href="working-with-digital-pathology-data-and-python.html#cb178-27" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>,<span class="dv">8</span>))</span>
<span id="cb178-28"><a href="working-with-digital-pathology-data-and-python.html#cb178-28" aria-hidden="true" tabindex="-1"></a>plt.imshow(slide_in_thumb_600_np)    </span>
<span id="cb178-29"><a href="working-with-digital-pathology-data-and-python.html#cb178-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-30"><a href="working-with-digital-pathology-data-and-python.html#cb178-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-31"><a href="working-with-digital-pathology-data-and-python.html#cb178-31" aria-hidden="true" tabindex="-1"></a><span class="co">#Get slide_in dims at each level. Remember that whole slide_in images store information</span></span>
<span id="cb178-32"><a href="working-with-digital-pathology-data-and-python.html#cb178-32" aria-hidden="true" tabindex="-1"></a><span class="co">#as pyramid at various levels</span></span>
<span id="cb178-33"><a href="working-with-digital-pathology-data-and-python.html#cb178-33" aria-hidden="true" tabindex="-1"></a>dims <span class="op">=</span> slide_in.level_dimensions</span>
<span id="cb178-34"><a href="working-with-digital-pathology-data-and-python.html#cb178-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-35"><a href="working-with-digital-pathology-data-and-python.html#cb178-35" aria-hidden="true" tabindex="-1"></a>num_levels <span class="op">=</span> <span class="bu">len</span>(dims)</span>
<span id="cb178-36"><a href="working-with-digital-pathology-data-and-python.html#cb178-36" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Number of levels in this image are:&quot;</span>, num_levels)</span>
<span id="cb178-37"><a href="working-with-digital-pathology-data-and-python.html#cb178-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-38"><a href="working-with-digital-pathology-data-and-python.html#cb178-38" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Dimensions of various levels in this image are:&quot;</span>, dims)</span>
<span id="cb178-39"><a href="working-with-digital-pathology-data-and-python.html#cb178-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-40"><a href="working-with-digital-pathology-data-and-python.html#cb178-40" aria-hidden="true" tabindex="-1"></a><span class="co">#By how much are levels downsampled from the original image?</span></span>
<span id="cb178-41"><a href="working-with-digital-pathology-data-and-python.html#cb178-41" aria-hidden="true" tabindex="-1"></a>factors <span class="op">=</span> slide_in.level_downsamples</span>
<span id="cb178-42"><a href="working-with-digital-pathology-data-and-python.html#cb178-42" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Each level is downsampled by an amount of: &quot;</span>, factors)</span>
<span id="cb178-43"><a href="working-with-digital-pathology-data-and-python.html#cb178-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-44"><a href="working-with-digital-pathology-data-and-python.html#cb178-44" aria-hidden="true" tabindex="-1"></a><span class="co">#Copy an image from a level</span></span>
<span id="cb178-45"><a href="working-with-digital-pathology-data-and-python.html#cb178-45" aria-hidden="true" tabindex="-1"></a>level3_dim <span class="op">=</span> dims[<span class="dv">2</span>]</span>
<span id="cb178-46"><a href="working-with-digital-pathology-data-and-python.html#cb178-46" aria-hidden="true" tabindex="-1"></a><span class="co">#Give pixel coordinates (top left pixel in the original large image)</span></span>
<span id="cb178-47"><a href="working-with-digital-pathology-data-and-python.html#cb178-47" aria-hidden="true" tabindex="-1"></a><span class="co">#Also give the level number (for level 3 we are providing a valueof 2)</span></span>
<span id="cb178-48"><a href="working-with-digital-pathology-data-and-python.html#cb178-48" aria-hidden="true" tabindex="-1"></a><span class="co">#Size of your output image</span></span>
<span id="cb178-49"><a href="working-with-digital-pathology-data-and-python.html#cb178-49" aria-hidden="true" tabindex="-1"></a><span class="co">#Remember that the output would be a RGBA image (Not, RGB)</span></span>
<span id="cb178-50"><a href="working-with-digital-pathology-data-and-python.html#cb178-50" aria-hidden="true" tabindex="-1"></a>level3_img <span class="op">=</span> slide_in.read_region((<span class="dv">0</span>,<span class="dv">0</span>), <span class="dv">2</span>, level3_dim) <span class="co">#Pillow object, mode=RGBA</span></span>
<span id="cb178-51"><a href="working-with-digital-pathology-data-and-python.html#cb178-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-52"><a href="working-with-digital-pathology-data-and-python.html#cb178-52" aria-hidden="true" tabindex="-1"></a><span class="co">#Convert the image to RGB</span></span>
<span id="cb178-53"><a href="working-with-digital-pathology-data-and-python.html#cb178-53" aria-hidden="true" tabindex="-1"></a>level3_img_RGB <span class="op">=</span> level3_img.convert(<span class="st">&#39;RGB&#39;</span>)</span>
<span id="cb178-54"><a href="working-with-digital-pathology-data-and-python.html#cb178-54" aria-hidden="true" tabindex="-1"></a>level3_img_RGB.show()</span>
<span id="cb178-55"><a href="working-with-digital-pathology-data-and-python.html#cb178-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-56"><a href="working-with-digital-pathology-data-and-python.html#cb178-56" aria-hidden="true" tabindex="-1"></a><span class="co">#Convert the image into numpy array for processing</span></span>
<span id="cb178-57"><a href="working-with-digital-pathology-data-and-python.html#cb178-57" aria-hidden="true" tabindex="-1"></a>level3_img_np <span class="op">=</span> np.array(level3_img_RGB)</span>
<span id="cb178-58"><a href="working-with-digital-pathology-data-and-python.html#cb178-58" aria-hidden="true" tabindex="-1"></a>plt.imshow(level3_img_np)</span>
<span id="cb178-59"><a href="working-with-digital-pathology-data-and-python.html#cb178-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-60"><a href="working-with-digital-pathology-data-and-python.html#cb178-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-61"><a href="working-with-digital-pathology-data-and-python.html#cb178-61" aria-hidden="true" tabindex="-1"></a><span class="co">#Return the best level for displaying the given downsample.</span></span>
<span id="cb178-62"><a href="working-with-digital-pathology-data-and-python.html#cb178-62" aria-hidden="true" tabindex="-1"></a>SCALE_FACTOR <span class="op">=</span> <span class="dv">32</span></span>
<span id="cb178-63"><a href="working-with-digital-pathology-data-and-python.html#cb178-63" aria-hidden="true" tabindex="-1"></a>best_level <span class="op">=</span> slide_in.get_best_level_for_downsample(SCALE_FACTOR)</span>
<span id="cb178-64"><a href="working-with-digital-pathology-data-and-python.html#cb178-64" aria-hidden="true" tabindex="-1"></a><span class="co">#Here it returns the best level to be 2 (third level)</span></span>
<span id="cb178-65"><a href="working-with-digital-pathology-data-and-python.html#cb178-65" aria-hidden="true" tabindex="-1"></a><span class="co">#If you change the scale factor to 2, it will suggest the best level to be 0 (our 1st level)</span></span>
<span id="cb178-66"><a href="working-with-digital-pathology-data-and-python.html#cb178-66" aria-hidden="true" tabindex="-1"></a><span class="co">#################################</span></span>
<span id="cb178-67"><a href="working-with-digital-pathology-data-and-python.html#cb178-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-68"><a href="working-with-digital-pathology-data-and-python.html#cb178-68" aria-hidden="true" tabindex="-1"></a><span class="co">#Generating tiles for deep learning training or other processing purposes</span></span>
<span id="cb178-69"><a href="working-with-digital-pathology-data-and-python.html#cb178-69" aria-hidden="true" tabindex="-1"></a><span class="co">#We can use read_region function and slide_in over the large image to extract tiles</span></span>
<span id="cb178-70"><a href="working-with-digital-pathology-data-and-python.html#cb178-70" aria-hidden="true" tabindex="-1"></a><span class="co">#but an easier approach would be to use DeepZoom based generator.</span></span>
<span id="cb178-71"><a href="working-with-digital-pathology-data-and-python.html#cb178-71" aria-hidden="true" tabindex="-1"></a><span class="co"># https://openslide.org/api/python/</span></span>
<span id="cb178-72"><a href="working-with-digital-pathology-data-and-python.html#cb178-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-73"><a href="working-with-digital-pathology-data-and-python.html#cb178-73" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> openslide.deepzoom <span class="im">import</span> DeepZoomGenerator</span>
<span id="cb178-74"><a href="working-with-digital-pathology-data-and-python.html#cb178-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-75"><a href="working-with-digital-pathology-data-and-python.html#cb178-75" aria-hidden="true" tabindex="-1"></a><span class="co">#Generate object for tiles using the DeepZoomGenerator</span></span>
<span id="cb178-76"><a href="working-with-digital-pathology-data-and-python.html#cb178-76" aria-hidden="true" tabindex="-1"></a>tiles <span class="op">=</span> DeepZoomGenerator(slide_in, tile_size<span class="op">=</span><span class="dv">256</span>, overlap<span class="op">=</span><span class="dv">0</span>, limit_bounds<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb178-77"><a href="working-with-digital-pathology-data-and-python.html#cb178-77" aria-hidden="true" tabindex="-1"></a><span class="co">#Here, we have divided our svs into tiles of size 256 with no overlap. </span></span>
<span id="cb178-78"><a href="working-with-digital-pathology-data-and-python.html#cb178-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-79"><a href="working-with-digital-pathology-data-and-python.html#cb178-79" aria-hidden="true" tabindex="-1"></a><span class="co">#The tiles object also contains data at many levels. </span></span>
<span id="cb178-80"><a href="working-with-digital-pathology-data-and-python.html#cb178-80" aria-hidden="true" tabindex="-1"></a><span class="co">#To check the number of levels</span></span>
<span id="cb178-81"><a href="working-with-digital-pathology-data-and-python.html#cb178-81" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;The number of levels in the tiles object are: &quot;</span>, tiles.level_count)</span>
<span id="cb178-82"><a href="working-with-digital-pathology-data-and-python.html#cb178-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-83"><a href="working-with-digital-pathology-data-and-python.html#cb178-83" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;The dimensions of data in each level are: &quot;</span>, tiles.level_dimensions)</span>
<span id="cb178-84"><a href="working-with-digital-pathology-data-and-python.html#cb178-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-85"><a href="working-with-digital-pathology-data-and-python.html#cb178-85" aria-hidden="true" tabindex="-1"></a><span class="co">#Total number of tiles in the tiles object</span></span>
<span id="cb178-86"><a href="working-with-digital-pathology-data-and-python.html#cb178-86" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Total number of tiles = : &quot;</span>, tiles.tile_count)</span>
<span id="cb178-87"><a href="working-with-digital-pathology-data-and-python.html#cb178-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-88"><a href="working-with-digital-pathology-data-and-python.html#cb178-88" aria-hidden="true" tabindex="-1"></a><span class="co">#How many tiles at a specific level?</span></span>
<span id="cb178-89"><a href="working-with-digital-pathology-data-and-python.html#cb178-89" aria-hidden="true" tabindex="-1"></a>level_num <span class="op">=</span> <span class="dv">11</span></span>
<span id="cb178-90"><a href="working-with-digital-pathology-data-and-python.html#cb178-90" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Tiles shape at level &quot;</span>, level_num, <span class="st">&quot; is: &quot;</span>, tiles.level_tiles[level_num])</span>
<span id="cb178-91"><a href="working-with-digital-pathology-data-and-python.html#cb178-91" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;This means there are &quot;</span>, tiles.level_tiles[level_num][<span class="dv">0</span>]<span class="op">*</span>tiles.level_tiles[level_num][<span class="dv">1</span>], <span class="st">&quot; total tiles in this level&quot;</span>)</span>
<span id="cb178-92"><a href="working-with-digital-pathology-data-and-python.html#cb178-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-93"><a href="working-with-digital-pathology-data-and-python.html#cb178-93" aria-hidden="true" tabindex="-1"></a><span class="co">#Dimensions of the tile (tile size) for a specific tile from a specific layer</span></span>
<span id="cb178-94"><a href="working-with-digital-pathology-data-and-python.html#cb178-94" aria-hidden="true" tabindex="-1"></a>tile_dims <span class="op">=</span> tiles.get_tile_dimensions(<span class="dv">11</span>, (<span class="dv">3</span>,<span class="dv">4</span>)) <span class="co">#Provide deep zoom level and address (column, row)</span></span>
<span id="cb178-95"><a href="working-with-digital-pathology-data-and-python.html#cb178-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-96"><a href="working-with-digital-pathology-data-and-python.html#cb178-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-97"><a href="working-with-digital-pathology-data-and-python.html#cb178-97" aria-hidden="true" tabindex="-1"></a><span class="co">#Tile count at the highest resolution level (level 16 in our tiles)</span></span>
<span id="cb178-98"><a href="working-with-digital-pathology-data-and-python.html#cb178-98" aria-hidden="true" tabindex="-1"></a>tile_count_in_large_image <span class="op">=</span> tiles.level_tiles[<span class="dv">16</span>] <span class="co">#126 x 151 (32001/256 = 126 with no overlap pixels)</span></span>
<span id="cb178-99"><a href="working-with-digital-pathology-data-and-python.html#cb178-99" aria-hidden="true" tabindex="-1"></a><span class="co">#Check tile size for some random tile</span></span>
<span id="cb178-100"><a href="working-with-digital-pathology-data-and-python.html#cb178-100" aria-hidden="true" tabindex="-1"></a>tile_dims <span class="op">=</span> tiles.get_tile_dimensions(<span class="dv">16</span>, (<span class="dv">120</span>,<span class="dv">140</span>))</span>
<span id="cb178-101"><a href="working-with-digital-pathology-data-and-python.html#cb178-101" aria-hidden="true" tabindex="-1"></a><span class="co">#Last tiles may not have full 256x256 dimensions as our large image is not exactly divisible by 256</span></span>
<span id="cb178-102"><a href="working-with-digital-pathology-data-and-python.html#cb178-102" aria-hidden="true" tabindex="-1"></a>tile_dims <span class="op">=</span> tiles.get_tile_dimensions(<span class="dv">16</span>, (<span class="dv">125</span>,<span class="dv">150</span>))</span>
<span id="cb178-103"><a href="working-with-digital-pathology-data-and-python.html#cb178-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-104"><a href="working-with-digital-pathology-data-and-python.html#cb178-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-105"><a href="working-with-digital-pathology-data-and-python.html#cb178-105" aria-hidden="true" tabindex="-1"></a>single_tile <span class="op">=</span> tiles.get_tile(<span class="dv">16</span>, (<span class="dv">62</span>, <span class="dv">70</span>)) <span class="co">#Provide deep zoom level and address (column, row)</span></span>
<span id="cb178-106"><a href="working-with-digital-pathology-data-and-python.html#cb178-106" aria-hidden="true" tabindex="-1"></a>single_tile_RGB <span class="op">=</span> single_tile.convert(<span class="st">&#39;RGB&#39;</span>)</span>
<span id="cb178-107"><a href="working-with-digital-pathology-data-and-python.html#cb178-107" aria-hidden="true" tabindex="-1"></a>single_tile_RGB.show()</span>
<span id="cb178-108"><a href="working-with-digital-pathology-data-and-python.html#cb178-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-109"><a href="working-with-digital-pathology-data-and-python.html#cb178-109" aria-hidden="true" tabindex="-1"></a><span class="co">###### Saving each tile to local directory</span></span>
<span id="cb178-110"><a href="working-with-digital-pathology-data-and-python.html#cb178-110" aria-hidden="true" tabindex="-1"></a>cols, rows <span class="op">=</span> tiles.level_tiles[<span class="dv">16</span>]</span>
<span id="cb178-111"><a href="working-with-digital-pathology-data-and-python.html#cb178-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb178-112"><a href="working-with-digital-pathology-data-and-python.html#cb178-112" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb178-113"><a href="working-with-digital-pathology-data-and-python.html#cb178-113" aria-hidden="true" tabindex="-1"></a>tile_dir <span class="op">=</span> <span class="st">&quot;images/saved_tiles/original_tiles/&quot;</span></span>
<span id="cb178-114"><a href="working-with-digital-pathology-data-and-python.html#cb178-114" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> row <span class="kw">in</span> <span class="bu">range</span>(rows):</span>
<span id="cb178-115"><a href="working-with-digital-pathology-data-and-python.html#cb178-115" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> col <span class="kw">in</span> <span class="bu">range</span>(cols):</span>
<span id="cb178-116"><a href="working-with-digital-pathology-data-and-python.html#cb178-116" aria-hidden="true" tabindex="-1"></a>        tile_name <span class="op">=</span> os.path.join(tile_dir, <span class="st">&#39;</span><span class="sc">%d</span><span class="st">_</span><span class="sc">%d</span><span class="st">&#39;</span> <span class="op">%</span> (col, row))</span>
<span id="cb178-117"><a href="working-with-digital-pathology-data-and-python.html#cb178-117" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">&quot;Now saving tile with title: &quot;</span>, tile_name)</span>
<span id="cb178-118"><a href="working-with-digital-pathology-data-and-python.html#cb178-118" aria-hidden="true" tabindex="-1"></a>        temp_tile <span class="op">=</span> tiles.get_tile(<span class="dv">16</span>, (col, row))</span>
<span id="cb178-119"><a href="working-with-digital-pathology-data-and-python.html#cb178-119" aria-hidden="true" tabindex="-1"></a>        temp_tile_RGB <span class="op">=</span> temp_tile.convert(<span class="st">&#39;RGB&#39;</span>)</span>
<span id="cb178-120"><a href="working-with-digital-pathology-data-and-python.html#cb178-120" aria-hidden="true" tabindex="-1"></a>        temp_tile_np <span class="op">=</span> np.array(temp_tile_RGB)</span>
<span id="cb178-121"><a href="working-with-digital-pathology-data-and-python.html#cb178-121" aria-hidden="true" tabindex="-1"></a>        plt.imsave(tile_name <span class="op">+</span> <span class="st">&quot;.png&quot;</span>, temp_tile_np)</span></code></pre></div>
</div>
<div id="using-python-and-r-together" class="section level2 hasAnchor" number="11.4">
<h2><span class="header-section-number">11.4</span> Using python and R together<a href="working-with-digital-pathology-data-and-python.html#using-python-and-r-together" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Using the <code>reticulate</code> package, R can communicate with python. This is slightly glitchy at the time of writing. So writing python objects from R seems to not work (using <code>py$</code> as mentioned in <code>reticulate</code> documentation), but python calling objects in the R environment seems reliable.</p>
<p>To do this we simply access the R environment in python using <code>r['myobject']</code>. Lets use the <code>file_path</code> example, where you might have used R to save the file path of specific slide file, then you want to use python to extract or manipulate the images.</p>
<p>In R:</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="working-with-digital-pathology-data-and-python.html#cb179-1" aria-hidden="true" tabindex="-1"></a>file_path <span class="ot">=</span> <span class="st">&#39;/home/test/slides/myslide.svs&#39;</span> <span class="co"># replace with where your slides are</span></span></code></pre></div>
<p>In python:</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb180-1"><a href="working-with-digital-pathology-data-and-python.html#cb180-1" aria-hidden="true" tabindex="-1"></a>file_path <span class="op">=</span> r[<span class="st">&#39;file_path&#39;</span>]</span>
<span id="cb180-2"><a href="working-with-digital-pathology-data-and-python.html#cb180-2" aria-hidden="true" tabindex="-1"></a>slide_in <span class="op">=</span> open_slide(file_path)</span></code></pre></div>
</div>
<div id="handling-errors-in-python" class="section level2 hasAnchor" number="11.5">
<h2><span class="header-section-number">11.5</span> Handling errors in python<a href="working-with-digital-pathology-data-and-python.html#handling-errors-in-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Most slides will be scanned well, newer scanners use lasers and a variety of different tools to automatically detect where tissue is and the depth at which to scan. However, some files will not scan properly or be corrupted in transfer (files are at least 200M+ per scan, ususally around 500M).</p>
<p>If there are errors, it is quite annoying, particularly given the size of files being handled if a run fails due to errors in opening. To avoid this we can use the python <code>try</code> commands, where it will attempt the command but continue on fail.</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb181-1"><a href="working-with-digital-pathology-data-and-python.html#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="co">#import pyvips</span></span>
<span id="cb181-2"><a href="working-with-digital-pathology-data-and-python.html#cb181-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> openslide <span class="im">import</span> open_slide</span>
<span id="cb181-3"><a href="working-with-digital-pathology-data-and-python.html#cb181-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> openslide</span>
<span id="cb181-4"><a href="working-with-digital-pathology-data-and-python.html#cb181-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> PIL <span class="im">import</span> Image</span>
<span id="cb181-5"><a href="working-with-digital-pathology-data-and-python.html#cb181-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb181-6"><a href="working-with-digital-pathology-data-and-python.html#cb181-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb181-7"><a href="working-with-digital-pathology-data-and-python.html#cb181-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tifffile <span class="im">as</span> tiff</span>
<span id="cb181-8"><a href="working-with-digital-pathology-data-and-python.html#cb181-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-9"><a href="working-with-digital-pathology-data-and-python.html#cb181-9" aria-hidden="true" tabindex="-1"></a><span class="co">#import</span></span>
<span id="cb181-10"><a href="working-with-digital-pathology-data-and-python.html#cb181-10" aria-hidden="true" tabindex="-1"></a>file_path <span class="op">=</span> r[<span class="st">&#39;file_path&#39;</span>]</span>
<span id="cb181-11"><a href="working-with-digital-pathology-data-and-python.html#cb181-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb181-12"><a href="working-with-digital-pathology-data-and-python.html#cb181-12" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb181-13"><a href="working-with-digital-pathology-data-and-python.html#cb181-13" aria-hidden="true" tabindex="-1"></a>  slide_in <span class="op">=</span> open_slide(file_path)</span>
<span id="cb181-14"><a href="working-with-digital-pathology-data-and-python.html#cb181-14" aria-hidden="true" tabindex="-1"></a>  slide_in_props <span class="op">=</span> slide_in.properties</span>
<span id="cb181-15"><a href="working-with-digital-pathology-data-and-python.html#cb181-15" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_properties&#39;</span>] <span class="op">=</span> slide_in_props</span>
<span id="cb181-16"><a href="working-with-digital-pathology-data-and-python.html#cb181-16" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_associated_images&#39;</span>] <span class="op">=</span> slide_in.associated_images.items()</span>
<span id="cb181-17"><a href="working-with-digital-pathology-data-and-python.html#cb181-17" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">Exception</span> <span class="im">as</span> e:</span>
<span id="cb181-18"><a href="working-with-digital-pathology-data-and-python.html#cb181-18" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_properties&#39;</span>] <span class="op">=</span> <span class="bu">print</span>(<span class="st">&quot;Error loading&quot;</span>) <span class="co"># doesn&#39;t save error</span></span>
<span id="cb181-19"><a href="working-with-digital-pathology-data-and-python.html#cb181-19" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_associated_images&#39;</span>] <span class="op">=</span> <span class="bu">print</span>(<span class="st">&quot;Error loading&quot;</span>) <span class="co"># doesn&#39;t save error</span></span></code></pre></div>
</div>
<div id="iterating-over-many-many-slide-files" class="section level2 hasAnchor" number="11.6">
<h2><span class="header-section-number">11.6</span> Iterating over many many slide files<a href="working-with-digital-pathology-data-and-python.html#iterating-over-many-many-slide-files" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Combining all the above, we can massively extend functionality and the number of slides we can process, for example to extract hundreds of slides at once and apply the same analysis to them.</p>
<p>In R:</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="working-with-digital-pathology-data-and-python.html#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb182-2"><a href="working-with-digital-pathology-data-and-python.html#cb182-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb182-3"><a href="working-with-digital-pathology-data-and-python.html#cb182-3" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">RETICULATE_PYTHON =</span> <span class="st">&quot;/home/slides/openslide_env/bin/python&quot;</span>)</span>
<span id="cb182-4"><a href="working-with-digital-pathology-data-and-python.html#cb182-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-5"><a href="working-with-digital-pathology-data-and-python.html#cb182-5" aria-hidden="true" tabindex="-1"></a><span class="co">#list files in a dir</span></span>
<span id="cb182-6"><a href="working-with-digital-pathology-data-and-python.html#cb182-6" aria-hidden="true" tabindex="-1"></a>files <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">file_path =</span> <span class="fu">list.files</span>(<span class="st">&#39;/home/slides&#39;</span>, </span>
<span id="cb182-7"><a href="working-with-digital-pathology-data-and-python.html#cb182-7" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">pattern =</span> <span class="st">&#39;.svs|.scn&#39;</span>,</span>
<span id="cb182-8"><a href="working-with-digital-pathology-data-and-python.html#cb182-8" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">full.names =</span> T, <span class="at">recursive =</span> T))<span class="co">#scn needs different</span></span>
<span id="cb182-9"><a href="working-with-digital-pathology-data-and-python.html#cb182-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-10"><a href="working-with-digital-pathology-data-and-python.html#cb182-10" aria-hidden="true" tabindex="-1"></a><span class="co">#make names</span></span>
<span id="cb182-11"><a href="working-with-digital-pathology-data-and-python.html#cb182-11" aria-hidden="true" tabindex="-1"></a>files <span class="ot">=</span> files <span class="sc">%&gt;%</span> </span>
<span id="cb182-12"><a href="working-with-digital-pathology-data-and-python.html#cb182-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">file_name =</span>  <span class="fu">basename</span>(file_path),</span>
<span id="cb182-13"><a href="working-with-digital-pathology-data-and-python.html#cb182-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">label_path =</span> <span class="fu">gsub</span>(<span class="st">&#39;.svs|.scn&#39;</span>, <span class="st">&#39;&#39;</span>, file_name),</span>
<span id="cb182-14"><a href="working-with-digital-pathology-data-and-python.html#cb182-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">out_path =</span> <span class="fu">gsub</span>(<span class="st">&#39;.svs|.scn&#39;</span>, <span class="st">&#39;_label.png&#39;</span>, file_path),</span>
<span id="cb182-15"><a href="working-with-digital-pathology-data-and-python.html#cb182-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">file_info =</span> <span class="fu">file.info</span>(file_path))</span>
<span id="cb182-16"><a href="working-with-digital-pathology-data-and-python.html#cb182-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-17"><a href="working-with-digital-pathology-data-and-python.html#cb182-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-18"><a href="working-with-digital-pathology-data-and-python.html#cb182-18" aria-hidden="true" tabindex="-1"></a><span class="co">#first get label properties</span></span>
<span id="cb182-19"><a href="working-with-digital-pathology-data-and-python.html#cb182-19" aria-hidden="true" tabindex="-1"></a>slide_properties_list <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb182-20"><a href="working-with-digital-pathology-data-and-python.html#cb182-20" aria-hidden="true" tabindex="-1"></a>associate_images_list <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb182-21"><a href="working-with-digital-pathology-data-and-python.html#cb182-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-22"><a href="working-with-digital-pathology-data-and-python.html#cb182-22" aria-hidden="true" tabindex="-1"></a><span class="co">#get properties</span></span>
<span id="cb182-23"><a href="working-with-digital-pathology-data-and-python.html#cb182-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-24"><a href="working-with-digital-pathology-data-and-python.html#cb182-24" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(files<span class="sc">$</span>label_path)){</span>
<span id="cb182-25"><a href="working-with-digital-pathology-data-and-python.html#cb182-25" aria-hidden="true" tabindex="-1"></a>  label_path <span class="ot">=</span> files<span class="sc">$</span>label_path[i]</span>
<span id="cb182-26"><a href="working-with-digital-pathology-data-and-python.html#cb182-26" aria-hidden="true" tabindex="-1"></a>  file_path <span class="ot">=</span> files<span class="sc">$</span>file_path[i]</span>
<span id="cb182-27"><a href="working-with-digital-pathology-data-and-python.html#cb182-27" aria-hidden="true" tabindex="-1"></a>  out_path <span class="ot">=</span> files<span class="sc">$</span>out_path[i]</span>
<span id="cb182-28"><a href="working-with-digital-pathology-data-and-python.html#cb182-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">source_python</span>(<span class="st">&#39;label_properties.py&#39;</span>, <span class="at">envir =</span> <span class="fu">globalenv</span>())</span>
<span id="cb182-29"><a href="working-with-digital-pathology-data-and-python.html#cb182-29" aria-hidden="true" tabindex="-1"></a>  slide_properties_list[[i]] <span class="ot">=</span> <span class="fu">as.character</span>(slide_properties)</span>
<span id="cb182-30"><a href="working-with-digital-pathology-data-and-python.html#cb182-30" aria-hidden="true" tabindex="-1"></a>  associate_images_list[[i]] <span class="ot">=</span> <span class="fu">as.character</span>(slide_associated_images)</span>
<span id="cb182-31"><a href="working-with-digital-pathology-data-and-python.html#cb182-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb182-32"><a href="working-with-digital-pathology-data-and-python.html#cb182-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb182-33"><a href="working-with-digital-pathology-data-and-python.html#cb182-33" aria-hidden="true" tabindex="-1"></a>openslide_data <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">do.call</span>(cbind, <span class="fu">list</span>(files<span class="sc">$</span>file_path, associate_images_list, slide_properties_list)))</span></code></pre></div>
<p>In python <code>label_properties.py</code>:</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb183-1"><a href="working-with-digital-pathology-data-and-python.html#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="co">#import pyvips</span></span>
<span id="cb183-2"><a href="working-with-digital-pathology-data-and-python.html#cb183-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> openslide <span class="im">import</span> open_slide</span>
<span id="cb183-3"><a href="working-with-digital-pathology-data-and-python.html#cb183-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> openslide</span>
<span id="cb183-4"><a href="working-with-digital-pathology-data-and-python.html#cb183-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> PIL <span class="im">import</span> Image</span>
<span id="cb183-5"><a href="working-with-digital-pathology-data-and-python.html#cb183-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb183-6"><a href="working-with-digital-pathology-data-and-python.html#cb183-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb183-7"><a href="working-with-digital-pathology-data-and-python.html#cb183-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tifffile <span class="im">as</span> tiff</span>
<span id="cb183-8"><a href="working-with-digital-pathology-data-and-python.html#cb183-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb183-9"><a href="working-with-digital-pathology-data-and-python.html#cb183-9" aria-hidden="true" tabindex="-1"></a><span class="co">#import</span></span>
<span id="cb183-10"><a href="working-with-digital-pathology-data-and-python.html#cb183-10" aria-hidden="true" tabindex="-1"></a>file_path <span class="op">=</span> r[<span class="st">&#39;file_path&#39;</span>]</span>
<span id="cb183-11"><a href="working-with-digital-pathology-data-and-python.html#cb183-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb183-12"><a href="working-with-digital-pathology-data-and-python.html#cb183-12" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb183-13"><a href="working-with-digital-pathology-data-and-python.html#cb183-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#slide_in = open_slide(&#39;/mnt/data/tdrake/light_microscopy/immune_staining_from_colin/01082019/collection_0000020061_2019-08-01 13_05_34.scn&#39;)</span></span>
<span id="cb183-14"><a href="working-with-digital-pathology-data-and-python.html#cb183-14" aria-hidden="true" tabindex="-1"></a>  slide_in <span class="op">=</span> open_slide(file_path)</span>
<span id="cb183-15"><a href="working-with-digital-pathology-data-and-python.html#cb183-15" aria-hidden="true" tabindex="-1"></a>  slide_in_props <span class="op">=</span> slide_in.properties</span>
<span id="cb183-16"><a href="working-with-digital-pathology-data-and-python.html#cb183-16" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_properties&#39;</span>] <span class="op">=</span> slide_in_props</span>
<span id="cb183-17"><a href="working-with-digital-pathology-data-and-python.html#cb183-17" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_associated_images&#39;</span>] <span class="op">=</span> slide_in.associated_images.items()</span>
<span id="cb183-18"><a href="working-with-digital-pathology-data-and-python.html#cb183-18" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">Exception</span> <span class="im">as</span> e:</span>
<span id="cb183-19"><a href="working-with-digital-pathology-data-and-python.html#cb183-19" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_properties&#39;</span>] <span class="op">=</span> <span class="bu">print</span>(<span class="st">&quot;Error loading&quot;</span>)</span>
<span id="cb183-20"><a href="working-with-digital-pathology-data-and-python.html#cb183-20" aria-hidden="true" tabindex="-1"></a>  r[<span class="st">&#39;slide_associated_images&#39;</span>] <span class="op">=</span> <span class="bu">print</span>(<span class="st">&quot;Error loading&quot;</span>)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="redcap.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="citations-with-r-markdown.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["cookbook.pdf", "cookbook.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
